storage/remote/otlptranslator/prometheusremotewrite/combined_appender.go

```diff
@@ -0,0 +1,159 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusremotewrite
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// NewCombinedAppender creates a combined appender that sets start times and
+// updates metadata for each series only once, and appends samples and
+// exemplars for each call.
+func NewCombinedAppender(app storage.Appender, logger *slog.Logger, reg prometheus.Registerer, ingestCTZeroSample bool) CombinedAppender {
+	return &combinedAppender{
+		app:                app,
+		logger:             logger,
+		ingestCTZeroSample: ingestCTZeroSample,
+		refs:               make(map[uint64]storage.SeriesRef),
+		samplesAppendedWithoutMetadata: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "api",
+			Name:      "otlp_without_metadata_appended_samples_total",
+			Help:      "The total number of received OTLP data points which were ingested without corresponding metadata.",
+		}),
+		outOfOrderExemplars: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "api",
+			Name:      "otlp_out_of_order_exemplars_total",
+			Help:      "The total number of received OTLP exemplars which were rejected because they were out of order.",
+		}),
+	}
+}
+
+// CombinedAppender is similar to storage.Appender, but combines updates to
+// metadata, created timestamps, exemplars and samples into a single call.
+type CombinedAppender interface {
+	// AppendSample appends a sample and related exemplars, metadata, and
+	// created timestamp to the storage.
+	AppendSample(ls labels.Labels, meta metadata.Metadata, t, ct int64, v float64, es []exemplar.Exemplar) error
+	// AppendSample appends a histogram and related exemplars, metadata, and
+	// created timestamp to the storage.
+	AppendHistogram(ls labels.Labels, meta metadata.Metadata, t, ct int64, h *histogram.Histogram, es []exemplar.Exemplar) error
+}
+
+type combinedAppender struct {
+	app                            storage.Appender
+	logger                         *slog.Logger
+	samplesAppendedWithoutMetadata prometheus.Counter
+	outOfOrderExemplars            prometheus.Counter
+	ingestCTZeroSample             bool
+	// Used to ensure we only update metadata and created timestamps once, and to share storage.SeriesRefs.
+	refs map[uint64]storage.SeriesRef
+}
+
+func (b *combinedAppender) AppendSample(ls labels.Labels, meta metadata.Metadata, t, ct int64, v float64, es []exemplar.Exemplar) (err error) {
```

Comment:
When adding the metadata to remote write, we need to use the metric family name and the old code simply set the name in the metadata to the OTEL name which doesn't have the magic _count, _bucket, _sum suffixes. However now I get the full name only, not the base name. And there will be metadata for individual series, not families, which confuses Grafana:
<img width="1288" height="157" alt="image" src="https://github.com/user-attachments/assets/bdaa36ef-a313-42e1-af60-b1da9d4a64bd" />

The interface should pass the base name as well, so we don't have to calculate it. For cases where the storage does not yet support per series metadata.

Reply:
Do we need it for AppendHistogram?

Reply:
Strictly speaking no, but I added it for consistency.